### PR TITLE
Initial commit of the vulkan backend

### PR DESCRIPTION
**Pull request by kkristof**

### CODE REVIEW

**Comment by kkristof** [comment]
It is a bit empty at the moment, but Rome wasn't built in a day :)

File: src/gpu/vulkan/gepard-vulkan.cpp [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+#include "gepard-vulkan.h"
+
+#include <iostream>
+
+namespace gepard {
+
+void GepardEngine::engineBackendInit()
+{
+    _engineBackend = reinterpret_cast<vulkan::GepardVulkan*>(new vulkan::GepardVulkan(_surface));
```
Comment (dbatyai) [inline src/gpu/vulkan/gepard-vulkan.cpp]: Why the `reinterpret_cast` here?

File: src/gpu/vulkan/gepard-vulkan.cpp [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+void GepardEngine::engineBackendDestroy()
+{
+    if (_engineBackend)
+        delete reinterpret_cast<vulkan::GepardVulkan*>(_engineBackend);
```
Comment (dbatyai) [inline src/gpu/vulkan/gepard-vulkan.cpp]: ditto

File: src/gepard-memory-buffer-surface.h [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+ */
+class MemoryBufferSurface : public Surface {
+public:
+    MemoryBufferSurface(uint32_t width = 0, uint32_t height = 0)
+        : Surface(width, height)
+    {
+        buffer = std::malloc(width * height * 4);
+    }
+
+    ~MemoryBufferSurface()
```
Comment (elecro) [inline src/gepard-memory-buffer-surface.h]: I think the destructors should be virtual.

Comment (kkristof) [inline src/gepard-memory-buffer-surface.h]: Okay

File: src/gpu/vulkan/gepard-vulkan.cpp (hunk 1 above)
Comment (kkristof) [inline src/gpu/vulkan/gepard-vulkan.cpp]: Same as in the GLES2 backend :)

File: src/gepard-memory-buffer-surface.h [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    {
+        if (buffer)
```
Comment (szledan) [inline src/gepard-memory-buffer-surface.h]: Please use brackets. (Only for early return.) ([CodeStyle](https://github.com/GepardGraphics/gepard/wiki/Coding-style))

File: src/gpu/vulkan/gepard-vulkan.cpp [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+namespace gepard {
+
+void GepardEngine::engineBackendInit()
+{
+    _engineBackend = reinterpret_cast<vulkan::GepardVulkan*>(new vulkan::GepardVulkan(_surface));
+}
+
+void GepardEngine::engineBackendDestroy()
+{
+    if (_engineBackend)
```
Comment (szledan) [inline src/gpu/vulkan/gepard-vulkan.cpp]: Please use brackets.

File: src/gepard-surface.h [hunk 6]
```diff
@@ -50,6 +50,7 @@ class Surface {
 
     virtual void* getDisplay() = 0;
     virtual unsigned long getWindow() = 0;
+	virtual void* getBuffer() = 0;
```
Comment (szledan) [inline src/gepard-surface.h]: tab -> space

File: src/gpu/vulkan/gepard-vulkan.cpp [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+void GepardVulkan::fillRect(float x, float y, float w, float h)
+{
+    std::cout << "fillrect: " << x << " " <<  x << " " <<  x << " " <<  x << std::endl;
+}
+
+void GepardVulkan::closePath()
+{
+}
+
+
```
Comment (szledan) [inline src/gpu/vulkan/gepard-vulkan.cpp]: extra line

**Comment by dbatyai** [comment]
LGTM

**Comment by szledan** [comment]
LGTM

**Pull request merged**